### PR TITLE
[Solved #17] Add schema validation with zod

### LINKED ISSUE #17: Schema validation
Epic:

### Tasks

- [x] Refactor socket data models
- [x] Research about schema validation
- [x] Scheme validation for the backend

### PR DESCRIPTION
**Pull request by supunTE**
Add schema validation using Zod and refactor sockets data models.
solved #17 

This PR is built on `supunTE:feature/7-write-unit-tests` and `supunTE:feature/18-generate-code-coverage`. Therefore It has many unmerged commits.

### CODE REVIEW

File: .github/workflows/onPR.yml [hunk 1]
```diff
@@ -20,3 +20,9 @@ jobs:
       - run: npm ci
       - run: npm run build 
       - run: npm run test
+      - run: npm run make-badges:ci
+      - run: npm run coverage:codecov
+      - name: Upload coverage to Codecov
+        uses: codecov/codecov-action@v2
+        with:
+          flags: coverage
```
Comment (pasinducw) [inline .github/workflows/onPR.yml]: Please add a new line character at the end of the file

File: .husky/pre-push [hunk 2]
```diff
@@ -2,3 +2,5 @@
 . "$(dirname -- "$0")/_/husky.sh"
 
 npm run test
+npm run make-badges
+git add README.md
```
Comment (pasinducw) [inline .husky/pre-push]: Please add a new line character at the end of the file

File: codecov.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+coverage:
+  status:
+    project:
+      default:
+        target: 70%
+
```
Comment (pasinducw) [inline codecov.yml]: Please have a new line character at the end of the file

File: libs/models/src/lib/sockets/player.ts [hunk 4]
```diff
@@ -25,18 +31,19 @@ export enum PlayerState {
   Racing = 'racing',
 }
 
+// ==== Types ==== //
+/** Player id template literal */
+export type playerId = z.input<typeof playerIdSchema>;
```
Comment (pasinducw) [inline libs/models/src/lib/sockets/player.ts]: Can we have this renamed to `PlayerId`

File: libs/models/src/lib/sockets/race.ts [hunk 5]
```diff
@@ -1,8 +1,20 @@
...
+import { PlayerLogsCollection } from './playerLog';
 
+// ==== Primary Schemas ==== //
+const raceIdSchema = z.custom<`T:${string}-R:${string}`>(id =>
+  /^T:[a-zA-Z0-9]{8}-R:[a-zA-Z0-9]{3}$/.test(id as string),
+);
+
+// ==== Types ==== //
+// Race id template literal
+export type raceId = z.input<typeof raceIdSchema>;
```
Comment (pasinducw) [inline libs/models/src/lib/sockets/race.ts]: Can we have this renamed to `RaceId`?

Comment (pasinducw) [reply libs/models/src/lib/sockets/race.ts]: Types need to be in PascalCase

File: libs/models/src/lib/sockets/tournament.ts [hunk 6]
```diff
@@ -1,7 +1,22 @@
...
 
+// ==== Primary Schemas ==== //
+export const timestampSchema = z
+  .number()
+  .int()
+  .nonnegative()
+  .refine(
+    value => value.toString().length === 13,
+    'Timestamp must be in milliseconds (13 digits).',
+  );
```
Comment (pasinducw) [inline libs/models/src/lib/sockets/tournament.ts:14]: Please insert an empty line

File: libs/models/src/lib/sockets/tournament.ts [hunk 7]
```diff
@@ -37,6 +52,12 @@ export enum TournamentState {
   Empty = 'empty',
 }
 
+// ==== Types ==== //
+/** Tournament id template literal */
+export type tournamentId = z.input<typeof tournamentIdSchema>;
```
Comment (pasinducw) [inline libs/models/src/lib/sockets/tournament.ts]: Please rename this to `TournamentId`. We need types to be in PsacalCase

File: libs/util/project.json [hunk 8]
```diff
@@ -23,12 +23,12 @@
     },
     "test": {
       "executor": "@nrwl/jest:jest",
-      "outputs": ["coverage/libs/util"],
+      "outputs": ["coverage"],
```
Comment (pasinducw) [inline libs/util/project.json]: Why is it like this here? I noticed that constants have a different path

**Review by pasinducw — CHANGES_REQUESTED** [review]

File: .github/workflows/onPR.yml (hunk 1 above)
Comment (supunTE) [reply .github/workflows/onPR.yml]: fixed in https://github.com/the-ai-team/razor/pull/19

Comment (supunTE) [reply .github/workflows/onPR.yml]: merged #19 into this

File: .husky/pre-push (hunk 2 above)
Comment (supunTE) [reply .husky/pre-push]: fixed

File: codecov.yml (hunk 3 above)
Comment (supunTE) [reply codecov.yml]: fixed

File: libs/models/src/lib/sockets/tournament.ts (hunk 7 above)
Comment (supunTE) [reply libs/models/src/lib/sockets/tournament.ts]: fixed in de1f40717ab164f66408e3bf2acbe282ff4ce7e7

File: libs/models/src/lib/sockets/race.ts (hunk 5 above)
Comment (supunTE) [reply libs/models/src/lib/sockets/race.ts]: fixed in de1f40717ab164f66408e3bf2acbe282ff4ce7e7

File: libs/models/src/lib/sockets/player.ts (hunk 4 above)
Comment (supunTE) [reply libs/models/src/lib/sockets/player.ts]: fixed de1f40717ab164f66408e3bf2acbe282ff4ce7e7

File: libs/util/project.json (hunk 8 above)
Comment (supunTE) [reply libs/util/project.json]: These output properties in project.json file are removed. coverage directories are already defined in jest.config.ts
fixed de1f40717ab164f66408e3bf2acbe282ff4ce7e7

File: libs/models/src/lib/sockets/tournament.ts (hunk 6 above)
Comment (supunTE) [reply libs/models/src/lib/sockets/tournament.ts:14]: fixed

File: libs/models/src/lib/sockets/tournament.ts (hunk 7 above)
Comment (pasinducw) [reply libs/models/src/lib/sockets/tournament.ts]: Wrong commit ref

File: libs/util/project.json (hunk 8 above)
Comment (pasinducw) [reply libs/util/project.json]: Wrong commit ref. Points to [style(data models/sockets): update naming convention of types](https://github.com/the-ai-team/razor/commit/de1f40717ab164f66408e3bf2acbe282ff4ce7e7)

Comment (supunTE) [reply libs/util/project.json]: refactor: remove unnecessary path defined for coverage [ecf12984f9c184cb98b3a1f916ff938d94279cc9]

**Review by dalanad — APPROVED** [review]

**Review by pasinducw — APPROVED** [review]

**Pull request merged**